src/main/java/frc/robot/LimitedMotor.java

```diff
@@ -0,0 +1,90 @@
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkMaxRelativeEncoder;
+
+class LimitedMotor {
+    // the motor being limited
+    private CANSparkMax spark;
+    // the encoder that belongs to the above spark
+    private RelativeEncoder encoder;
+
+    // the position of the encoder in rotations when it hits the reverse limit switch.
+    private double start;
+    // the distance in rotations from the forward to the reverse limit switch.
+    private final double range;
+    // the maximum speed, in RPM, at which it is safe to run into the limit switch
+    private final double safeSpeed;
+
+    // Tells us whether we have found where the reverse limit switch is
+    private boolean startIsKnown = false;
+
+    // where we want it to go to
+    // Where 1 is the forward limit switch and 0 is the reverse limit switch
+    private double targetPos;
+
+    /**
+     * Creates a new limited motor
+     * 
+     * @param spark the motor to be limited
+     * @param range the distance in rotations from the forward to the reverse limit switch
+     * @param safeSpeed the maximum speed, in RPM, at which it is safe to run into the limit switch
+     */
+    public LimitedMotor(CANSparkMax spark, double range, double safeSpeed) {
+        this.spark = spark;
+        this.range = range;
+        this.safeSpeed = safeSpeed;
+        // 4096 is the ticks per revolution of the encoder
+        this.encoder = spark.getEncoder(SparkMaxRelativeEncoder.Type.kQuadrature, 4096);
+    }
+
+    /**
+     * Creates a new limited motor
+     * 
+     * @param spark the motor to be limited
+     * @param range the distance in rotations from the forward to the reverse limit switch
+     * @param safeSpeed the maximum speed, in RPM, at which it is safe to run into the limit switch
+     */
+    public LimitedMotor(IMotorController motorController, double range, double safeSpeed) {
+        if (!(motorController instanceof SparkMotorController)) {
+            throw new IllegalArgumentException("Limited motor must be a spark motor controller");
+        }
+        CANSparkMax spark = ((SparkMotorController)motorController).controller;
+        LimitedMotor(spark, range, safeSpeed);
+    }
+
+    /**
+     * this method should be called in robot periodic
+     * moves the motor closer to the target position
+     */
+    public void tick() {
+        
+        if (!this.startIsKnown){
```

Comment:
Is it necessary to use `this` everywhere? My personal codestyle would say to remove it, but it may be good for readability and avoiding possible conflicts between properties in the class and variables in the method.